requests/models.py

```diff
@@ -365,11 +365,19 @@ def prepare_url(self, url, params):
         if not host:
             raise InvalidURL("Invalid URL %r: No host supplied" % url)
 
-        # Only want to apply IDNA to the hostname
-        try:
-            host = idna.encode(host, uts46=True).decode('utf-8')
-        except (UnicodeError, idna.IDNAError):
-            raise InvalidURL('URL has an invalid label.')
+        # In general, we want to IDNA encode everything, as that allows users
+        # to automatically get correct behaviour in the majority of cases.
+        # However, we’re quite strict about IDNA encoding, so if a user has
+        # already IDNA encoded their domain name, we shouldn’t bother validating
+        # their work. We determine something *looks* IDNA encoded by checking
+        # for the IDNA-encoding tag (xn--) at the start of the hostname, and
+        # confirming that the hostname is entirely ASCII characters.
+        if not host.startswith('xn--') or not is_ascii(host):
```

Comment:
Let's try to avoid problems by setting this to `u'xn--'`. On top of this, given that we know that the string will always be unicode here, we can safely rewrite `is_ascii` to `unicode_is_ascii` which will just work for unicode strings for now.

Save ourselves the unneeded complexity.